tower-steer/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+//!
+//!     let reqs = vec!["A", "b", "C", "d"];
+//!     let reqs: Vec<String> = reqs.into_iter().map(String::from).collect();
+//!     for r in reqs {
+//!         poll_fn(|cx| s.poll_ready(cx)).await.unwrap();
+//!         s.call(r).await;
+//!     }
+//! }
+//! ```
+#![deny(missing_docs)]
+#![warn(unreachable_pub)]
+#![warn(missing_copy_implementations)]
+#![warn(trivial_casts)]
+#![warn(trivial_numeric_casts)]
+#![warn(unused_extern_crates)]
+#![warn(rust_2018_idioms)]
+#![warn(missing_debug_implementations)]
+#![allow(clippy::type_complexity)]
+
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower_service::Service;
+
+type StdError = Box<dyn std::error::Error + Send + Sync + 'static>;
+
+/// This is how callers of [`Steer`] tell it which `Service` a `Req` corresponds to.
+pub trait Picker<S, Req> {
+    /// Return an index into the iterator of `Service` passed to [`Steer::new`].
+    fn pick(&mut self, r: &Req, services: &[S]) -> usize;
+}
+
+impl<S, F, Req> Picker<S, Req> for F
+where
+    F: Fn(&Req, &[S]) -> usize,
+{
+    fn pick(&mut self, r: &Req, services: &[S]) -> usize {
+        self(r, services)
+    }
+}
+
+/// `Steer` manages a list of `Service`s which all handle the same type of request.
+///
+/// An example use case is a sharded service.
+/// It accepts new requests, then:
+/// 1. Determines, via the provided [`Picker`], which `Service` the request coresponds to.
+/// 2. Waits (in `poll_ready`) for *all* services to be ready.
+/// 3. Calls the correct `Service` with the request, and returns a future corresponding to the
+///    call.
+#[derive(Debug)]
+pub struct Steer<S, F, Req> {
+    router: F,
+    // tuple of is_ready, service
+    cls: Vec<S>,
+    ready: Vec<bool>,
+    _phantom: std::marker::PhantomData<Req>,
+}
+
+impl<S, F, Req> Steer<S, F, Req>
+where
+    S: Service<Req, Error = StdError>,
+    S::Future: 'static,
+{
+    /// Make a new [`Steer`] with a list of `Service`s and a `Picker`.
+    ///
+    /// Note: the order of the `Service`s is significant for [`Picker::pick`]'s return value.
+    pub fn new(cls: impl IntoIterator<Item = S>, router: F) -> Self {
+        let cls: Vec<_> = cls.into_iter().collect();
+        let ready: Vec<_> = cls.iter().map(|_| false).collect();
+        Self {
+            router,
+            cls,
+            ready,
+            _phantom: Default::default(),
+        }
+    }
+}
+
+impl<S, Req, T, F> Service<Req> for Steer<S, F, Req>
+where
+    S: Service<Req, Response = T, Error = StdError>,
+    S::Future: 'static,
+    F: Picker<S, Req>,
+{
+    type Response = T;
+    type Error = StdError;
+    type Future = Pin<Box<dyn Future<Output = Result<T, StdError>>>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        use futures_util::ready;
+        // must wait for *all* services to be ready.
+        // this will cause head-of-line blocking unless the underlying services are always ready.
```

Comment:
May be useful to include this in the docs somewhere?

Reply:
it is: https://github.com/tower-rs/tower/pull/426/files/1cde2511a0fdee7e56a6f9ee4327edb74fcb1bd2#diff-91249aa82a3e4b3ca98888fc272df30dR86

maybe it can be made more prominent

Reply:
Ah, sorry, no, I was referring specifically to the advice about using an unbounded channel in front of each service / making sure it always returns `poll_ready`.

Reply:
yeah, this is added to docs in the latest commit